### ISSUE TITLE
dotnet restore nor nuget restore CLI command evalute the Directory.Build.props file

### ISSUE THREAD

**Issue by stgerbb** [NuGet/Home#12442]
### Describe the bug
When running the `dotnet restore` or `nuget restore` CLI command then defined properties in the _Directory.Build.props_ file are not evaluated. Only the csproj is taken into account. This causes properties to be duplicated from the _Directory.Build.props_ file to the csproj.

### To Reproduce
1. Take a .nupkg file and name it the following: _Debug.Test.nupkg_.
2. Place it in a local feed on the PC where Visual Studio can find it.
3. Create a new C# project (i.e. a class library)
4. Create a _Directory.Build.props_ file and place it next to the csproj file.
5. Add the following to the _Directory.Build.props_ file:
```
<PropertyGroup Condition="$(Configuration.StartsWith('Debug'))">
	<PureConfiguration>Debug</PureConfiguration>
</PropertyGroup>
```
6. Reference the Debug.Test NuGet in the previous created project the following way: 
`<PackageReference Include="$(PureConfiguration).Test" Version="<TheVersionOfThePackage>" />`
7. Clear nuget cache
8. Run either `nuget restore NameOfTheProject.csproj` or `dotnet restore NameOfTheProject.csproj` from the command line (i.e. cmd.exe)

**Expected**: PackageReference should get resolved as _Debug.Test_, because the `$(PureConfiguration)` property is defined in the _Directory.Build.props file_.

**Result**: PackageReference is resolved as _.Test_, because the `$(PureConfiguration)` property was not found. Causes `dotnet restore` or `nuget restore` to fail to resolve the project.

**Workaround**: Duplicate the property from the _Directory.Build.props_ file into the csproj.

### Further technical details
dotnet --info
```
.NET SDK:
 Version:   7.0.102
 Commit:    4bbdd14480

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.102\

Host:
  Version:      7.0.2
  Architecture: x64
  Commit:       d037e070eb

.NET SDKs installed:
  5.0.400 [C:\Program Files\dotnet\sdk]
  5.0.401 [C:\Program Files\dotnet\sdk]
  5.0.403 [C:\Program Files\dotnet\sdk]
  5.0.406 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  7.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.19 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  arm64 [C:\Program Files\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\arm64\InstallLocation]
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

Running Visual Studio 2022 Version 17.4.4
Running nuget.exe v6.3.1
```

**Comment by zivkan** [comment]
TL;DR NuGet runs as a MSBuild task, and as such is subject to how MSBuild works. What you think is a bug is how MSBuild works.

@stgerbb MSBuild has docs on [how MSBuild builds projects](https://learn.microsoft.com/en-us/visualstudio/msbuild/build-process-overview?view=vs-2022). Unfortunately it's not explicit in what I want to point out, but it runs the program top-down (MSBuild is a scripting language, it's not a declarative file that allows arbitrary lookups, like Excel). Therefore, trying reference a property before it is defined will always return an empty string.

In case you're not familiar with MSBuild debugging, the most useful and powerful are binary logs, which you can get by adding the `-bl` argument to any MSBuild command. For example `dotnet build -bl`, or `dotnet restore -bl`. Unless you provide a path and filename to `-bl` to use, it by default will write `msbuild.binlog` to the current directory, which you can then use https://msbuildlog.com to view (there's an online viewer, or an app you can download.

In this specific example, in the MSBuild binlog viewer, right click your project file and select "Preprocess", but another way to get the same output is by running `dotnet msbuild -pp:pp.txt` on the command line, and `pp.txt` will contain the same "preprocessed" output. This evaluates all the MSBuild imports, so you can see exactly in what order different files (and therefore different properties) are evaluated.

When I try this with your example, the `<PureConfiguration>` property is defined on line 79 of the preprocessed file, whereas the [.NET SDK sets the default Configuration](https://github.com/dotnet/sdk/blob/3d71e0074b6f015a83460b6d0d6fe7ca4e7f8b32/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props#L23) on line 271. So, as I mentioned above, when trying to reference a property before it's defined, the value you'll get is always an empty string. You need to write your MSBuild scripts in a way that doesn't try to referene properties before they're defined.

While `dotnet restore` doesn't support the `-c` argument, build does an implicit restore, so we can use `dotnet build -c Debug` and verify that it works as you expect, since `Configuration` is now what MSBuild calls a global property. A similar outcome can be achieved with `dotnet restore -p:Configuration=Debug`.
